### PR TITLE
Add storage sink consumer developing guide

### PR DESCRIPTION
**Pull request by zhaoxinyu**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.6 (TiDB 6.6 versions)
- [x] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+// Query newly added files and the latest checkpoint from storage，one file can only be returned once
+func (c *Consumer) ExposeNewFiles() (int64, []string) {}
+
+
+// ConsumerManager is responsible for assigning tasks to TableConsumer.
+// Different consumers can consume data concurrently,
+// but data of one table must be processed by the same TableConsumer.
+type ConsumerManager struct {
+  // StorageCheckpoint indicates that the data whose transaction commit time is less than this checkpoint has been stored in storage
+  StorageCheckpoint int64
```
Comment (zhaoxinyu) [inline ticdc/ticdc-storage-consumer-dev-guide.md:41]: StorageCheckpoint 这个值从哪里计算出来，好像没有相关描述 @IANTHEREAL

Comment (IANTHEREAL) [reply ticdc/ticdc-storage-consumer-dev-guide.md:41]: 这个是从 storage 取的 checkpoint，可以增加一句注释

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+以下是 Consumer 消费路程中的组件和功能定义，及其功能注释:
+
+```
+type StorageReader struct {
+}
+
+// Read the files from storage
+// Add newly added files and delete files that not exist in storage
+func (c *Consumer) ReadFiles() {}
```
Comment (IANTHEREAL) [inline ticdc/ticdc-storage-consumer-dev-guide.md:30]:
```suggestion
func (c * StorageReader) ReadFiles() {}
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+// Query newly added files and the latest checkpoint from storage，one file can only be returned once
+func (c *Consumer) ExposeNewFiles() (int64, []string) {}
```
Comment (IANTHEREAL) [inline ticdc/ticdc-storage-consumer-dev-guide.md:33]:
```suggestion
func (c * StorageReader) ExposeNewFiles() (int64, []string) {}
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+
+// ConsumerManager is responsible for assigning tasks to TableConsumer.
+// Different consumers can consume data concurrently,
+// but data of one table must be processed by the same TableConsumer.
+type ConsumerManager struct {
+  // StorageCheckpoint indicates that the data whose transaction commit time is less than this checkpoint has been stored in storage
```
Comment (IANTHEREAL) [inline ticdc/ticdc-storage-consumer-dev-guide.md:40]:
```suggestion
  // StorageCheckpoint is fetched from storage
  // It indicates that the data whose transaction commit time is less than this checkpoint has been stored in storage
 
```

File: ticdc/ticdc-storage-consumer-dev-guide.md (hunk 1 above)
Comment (zhaoxinyu) [reply ticdc/ticdc-storage-consumer-dev-guide.md:41]: 嗯，可以改成从 storage 中保存的 metadata 文件里获取 checkpoint

**Comment by IANTHEREAL** [comment]
Rest LGTM

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    │       └── schema.json
+```
+
+此时需要先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句，分为三种情况处理，接着再开始同步数据文件 `CDC000001.json`:
+
+- 如果 Query 为空则忽略；
+- 如果 TableVersion 小于 consumer checkpoint 则忽略；
+- 否则，执行到下游 MySQL。
+
+例如 `test/tbl_1/437752935075545091/schema.json` 文件内容如下：
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:121]: 需要调整，参考：如下是一个 Query DDL 结果不为空的例子：

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    │   └── 437752935075546092
+    │   │   └── CDC000001.json
+    │   │   └── schema.json
+```
+
+消费逻辑跟上述一致，先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句，分为前文中提到的三种情况处理，接着再开始同步数据文件 `CDC000001.json`。
+
+## 数据事件的处理
+
+根据 `{schema}/{table}/{table-version-separator}/schema.json` 文件处理好 DDL 事件后，就可以在 `{schema}/{table}/{table-version-separator}/` 目录下，根据具体的文件格式并按照文件序号依次处理数据变更事件。
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:179]: 修改标题，并调整 Line 179

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+```
+├── metadata
+└── test
+    ├── tbl_1
+    │   └── 437752935075545091
+    │       ├── CDC000001.json
+    │       └── schema.json
+```
+
+此时需要先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句，分为三种情况处理，接着再开始同步数据文件 `CDC000001.json`:
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:115]: 这句需要移到后面：
> 接着再开始同步数据文件 `CDC000001.json`

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+TiCDC 不提供消费存储服务的数据的标准实现，以下介绍一个 Golang 版本的消费示例程序，该示例程序会读取存储服务中的数据并写入到下游 MySQL 兼容数据库。你可以参考本文档提供的数据格式和以下例子实现消费端。
+
+[Golang 例子](https://github.com/pingcap/tiflow/tree/master/cmd/storage-consumer)
+
+## Consumer 设计
+
+下图是 Consumer 的整体消费流程：
+![TiCDC storage consumer overview](/media/ticdc/ticdc-storage-consumer-overview.png)
+
+以下是 Consumer 消费路程中的组件和功能定义，及其功能注释:
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:22]:
```suggestion
以下是 Consumer 消费流程中的组件和功能定义，及其功能注释:
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
+---
+title: Storage sink 消费程序开发指引
+summary: 了解如何设计与实现一个 storage sink 的消费程序。
+---
+
+# Storage sink 消费程序设计
+
+本文介绍如何设计和实现一个 TiDB 数据变更的消费程序。
+
+> **注意：**
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:10]:
```suggestion
> **注意：**
>
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+> 当前 Storage sink 无法处理 `DROP DATABASE` DDL，请你尽量避免执行该语句。如果你需要执行 `DROP DATABASE` DDL，请你在下游 MySQL 手动执行。
+
+TiCDC 不提供消费存储服务的数据的标准实现，以下介绍一个 Golang 版本的消费示例程序，该示例程序会读取存储服务中的数据并写入到下游 MySQL 兼容数据库。你可以参考本文档提供的数据格式和以下例子实现消费端。
+
+[Golang 例子](https://github.com/pingcap/tiflow/tree/master/cmd/storage-consumer)
+
+## Consumer 设计
+
+下图是 Consumer 的整体消费流程：
+![TiCDC storage consumer overview](/media/ticdc/ticdc-storage-consumer-overview.png)
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:21]:
```suggestion
下图是 Consumer 的整体消费流程：

![TiCDC storage consumer overview](/media/ticdc/ticdc-storage-consumer-overview.png)
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    │   └── 437752935075546092
+    │   │   └── CDC000001.json
+    │   │   └── schema.json
+```
+
+消费逻辑跟上述一致，先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句，分为前文中提到的三种情况处理，接着再开始同步数据文件 `CDC000001.json`。
+
+## DML 事件的处理
+
+处理好 DDL 事件后，就可以在 `{schema}/{table}/{table-version-separator}/` 目录下，根据具体的文件格式（CSV/Canal-JSON）并按照文件序号依次处理 DML 事件。
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:182]:
```suggestion
处理好 DDL 事件后，就可以在 `{schema}/{table}/{table-version-separator}/` 目录下，根据具体的文件格式（CSV 或 Canal-JSON）并按照文件序号依次处理 DML 事件。
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    ├── tbl_1
+    │   ├── 437752935075545091
+    │   │   ├── CDC000001.json
+    │   │   └── schema.json
+    │   └── 437752935075546092
+    │   │   └── CDC000001.json
+    │   │   └── schema.json
+```
+
+消费逻辑跟上述一致，先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句，分为前文中提到的三种情况处理，接着再开始同步数据文件 `CDC000001.json`。
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:178]:
```suggestion
消费逻辑跟上述一致，先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句并按不同情况处理，然后同步数据文件 `CDC000001.json`。
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+            "ColumnName":"OfficeLocation",
+            "ColumnType":"BLOB",
+            "ColumnLength":"20"
+        }
+    ],
+    "TableColumnsTotal":"5"
+}
+```
+
+然后当程序又一次遍历目录，发现该表多出了一个新的版本目录。注意要在 `test/tbl_1/437752935075545091` 目录下所有文件消费完成后，然后才能开始消费新的目录下的数据。
```
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:164]:
```suggestion
当程序再次遍历目录，发现该表新增了一个版本目录。程序先消费完 `test/tbl_1/437752935075545091` 目录下的所有文件，然后再消费新目录下的数据。
```

File: ticdc/ticdc-storage-consumer-dev-guide.md (hunk 12 above)
Comment (shichun-0415) [inline ticdc/ticdc-storage-consumer-dev-guide.md:178]:
```suggestion
消费逻辑跟上述一致，先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句并按不同情况处理，然后同步数据文件 `CDC000001.json`。
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+本文介绍如何设计和实现一个 TiDB 数据变更的消费程序。
+
+> **注意：**
+> 
+> 当前 Storage sink 无法处理 `DROP DATABASE` DDL，请你尽量避免执行该语句。如果你需要执行 `DROP DATABASE` DDL，请你在下游 MySQL 手动执行。
+
+TiCDC 不提供消费存储服务的数据的标准实现，以下介绍一个 Golang 版本的消费示例程序，该示例程序会读取存储服务中的数据并写入到下游 MySQL 兼容数据库。你可以参考本文档提供的数据格式和以下例子实现消费端。
+
+[Golang 例子](https://github.com/pingcap/tiflow/tree/master/cmd/storage-consumer)
```
Comment (Oreoxmt) [inline ticdc/ticdc-storage-consumer-dev-guide.md:16]:
```suggestion
[Golang 示例代码](https://github.com/pingcap/tiflow/tree/master/cmd/storage-consumer)
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+    ├── tbl_1
+    │   └── 437752935075545091
+    │       ├── CDC000001.json
+    │       └── schema.json
+```
+
+此时需要先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句，分为两种情况处理:
+
+- 如果 Query 为空或者 TableVersion 小于 consumer checkpoint 则跳过；
+- 否则，在下游 MySQL 执行获取到的 DDL 语句。
```
Comment (Oreoxmt) [inline ticdc/ticdc-storage-consumer-dev-guide.md:120]:
```suggestion
- 如果 DDL Query 为空或者 TableVersion 小于 consumer checkpoint 则跳过该语句。
- 否则，在下游 MySQL 数据库中执行获取到的 DDL 语句。
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+    ├── tbl_1
+    │   ├── 437752935075545091
+    │   │   ├── CDC000001.json
+    │   │   └── schema.json
+    │   └── 437752935075546092
+    │   │   └── CDC000001.json
+    │   │   └── schema.json
+```
+
+消费逻辑跟上述一致，先解析 `schema.json` 文件中的表结构信息，从中获取 DDL query 语句并按不同情况处理，然后同步数据文件 `CDC000001.json`。
```
Comment (Oreoxmt) [inline ticdc/ticdc-storage-consumer-dev-guide.md:178]:
```suggestion
消费逻辑跟上述一致，先解析 `schema.json` 文件中的表结构信息，从中获取 DDL Query 语句并按不同情况处理，然后同步数据文件 `CDC000001.json`。
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+## DML 事件的处理
+
+处理好 DDL 事件后，就可以在 `{schema}/{table}/{table-version-separator}/` 目录下，根据具体的文件格式（CSV 或 Canal-JSON）并按照文件序号依次处理 DML 事件。
```
Comment (Oreoxmt) [inline ticdc/ticdc-storage-consumer-dev-guide.md:182]:
```suggestion
在处理完 DDL 事件后，可以在 `{schema}/{table}/{table-version-separator}/` 目录下，根据具体的文件格式（CSV 或 Canal-JSON）并按照文件序号依次处理 DML 事件。
```

File: ticdc/ticdc-storage-consumer-dev-guide.md [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+
+因为 TiCDC 提供 At Least Once 语义，可能出现重复发送数据的情况，所以需要在消费程序中对比数据事件的 commit ts 和 consumer checkpoint，如果 commit ts 小于 consumer checkpoint 则需要做去重处理。
```
Comment (Oreoxmt) [inline ticdc/ticdc-storage-consumer-dev-guide.md:184]:
```suggestion
因为 TiCDC 提供 At Least Once 语义，可能出现重复发送数据的情况，所以需要在消费程序中比较数据事件的 commit ts 和 consumer checkpoint，并在 commit ts 小于 consumer checkpoint 的情况下进行去重处理。
```

**Review by Oreoxmt — COMMENTED** [review]
Rest LGTM

**Review by Oreoxmt — APPROVED** [review]
LGTM

**Review by shichun-0415 — APPROVED** [review]

**Pull request merged**